metagpt/provider/bedrock_api.py

```diff
@@ -0,0 +1,141 @@
+from typing import Literal
+import json
+from metagpt.const import USE_CONFIG_TIMEOUT
+from metagpt.provider.llm_provider_registry import register_provider
+from metagpt.configs.llm_config import LLMConfig, LLMType
+from metagpt.provider.base_llm import BaseLLM
+from metagpt.logs import log_llm_stream, logger
+from metagpt.provider.bedrock.bedrock_provider import get_provider
+from metagpt.provider.bedrock.utils import NOT_SUUPORT_STREAM_MODELS, get_max_tokens
+import boto3
+from botocore.eventstream import EventStream
+
+
+@register_provider([LLMType.BEDROCK])
+class BedrockLLM(BaseLLM):
+    def __init__(self, config: LLMConfig):
+        self.config = config
+        self.__client = self.__init_client("bedrock-runtime")
+        self.__provider = get_provider(self.config.model)
+        logger.warning(
+            "Amazon bedrock doesn't support asynchronous now")
+
+    def __init_client(self, service_name: Literal["bedrock-runtime", "bedrock"]):
+        """initialize boto3 client"""
+        # access key and secret key from https://us-east-1.console.aws.amazon.com/iam
+        self.__credentital_kwargs = {
+            "aws_secret_access_key": self.config.secret_key,
+            "aws_access_key_id": self.config.access_key,
+            "region_name": self.config.region_name
+        }
+        session = boto3.Session(**self.__credentital_kwargs)
+        client = session.client(service_name)
+        return client
+
+    @property
+    def client(self):
+        return self.__client
+
+    @property
+    def provider(self):
+        return self.__provider
+
+    def list_models(self):
+        """list all available text-generation models
+
+        ```shell
+        ai21.j2-ultra-v1                    Support Streaming:False
+        meta.llama3-70b-instruct-v1:0       Support Streaming:True
+        ……
+        ```
+        """
+        client = self.__init_client("bedrock")
+        # only output text-generation models
+        response = client.list_foundation_models(byOutputModality="TEXT")
+        summaries = [f'{summary["modelId"]:50} Support Streaming:{summary["responseStreamingSupported"]}'
+                     for summary in response["modelSummaries"]]
+        logger.info("\n"+"\n".join(summaries))
+
+    def invoke_model(self, request_body: str) -> dict:
+        response = self.__client.invoke_model(
+            modelId=self.config.model, body=request_body
+        )
+        usage = self._get_usage(response)
+        self._update_costs(usage)
+        response_body = self._get_response_body(response)
+        return response_body
+
+    def invoke_model_with_response_stream(self, request_body: str) -> EventStream:
+        response = self.__client.invoke_model_with_response_stream(
+            modelId=self.config.model, body=request_body)
+        usage = self._get_usage(response)
+        self._update_costs(usage)
```

Comment:
Here doesn't have the price of each 1k tokens for model, so `_update_costs` will fail. Don't you notice this in your log?